README.md
# Introduction

This is fork with datastore cache for http-responses. To enable it you must set env var DATASTORE_USE_LOCAL_CACHE=true.

The Microgateway is an developer-focused, extensible gateway framework written
in Node.js for enforcing access to Microservices & APIs - https://developer.ibm.com/apiconnect/. It supports the
following core features:

*	Secure and control access to APIs using Swagger (OpenAPI) Specification
*	Collection of pre-built gateway policies for API Key validation, OAuth 2.0,
    rate limiting, and JavaScript
*	Create gateway policies (security, routing, integration, etc... ) using
    Swagger extensions (API Assembly)
*	Simple interfaces for creating your own gateway policies.

The role of a Gateway in an API architecture is to protect, enrich and
control access to API services. These sets of capabilities are often related
to security and rate limiting, but it also includes the ability to do deeper
message inspection.  For example, you may want to insure that the message
received is properly formed JSON, XML, or data following your own specific
format.  In addition, the Gateway can modify the payload or transform it to
meet old or new interfaces for the API backend. Finally, the Gateway can
invoke multiple services and aggregate responses from multiple API backends.

The Microgateway is the foundation for all of those things. It is optimized
to perform security, rate limiting, and much more complex packet processing
through a highly flexible flow-engine.

The flow-engine is a processing unit that allows you to define a sequence
of processing policies or tasks that get applied to the transaction as it
passes through the Gateway for a specific API. These policies may be organized
in a linear fashion, executing in sequential order. Or, they may have
conditional logic deciding which set of policies will execute depending on
variables that are part of the API flow.

The Microgateway currently contains the following policies that run implicitly
based on swagger definitions:

* **Client_ID/Client_Secret** – Use a client ID and client secret to
  authenticate and authorize the use of this API
* **Basic Auth** – Use Basic Auth to authenticate and authorize the use of this
  API
* **OAuth 2.0** – Use OAuth2.0 to authenticate and authorize the use of this API
* **Rate-Limit** – Limit the number of requests per time unit that the
  subscriber may access this API

The Microgateway currently contains the following polices that are available
to the Assembly extension to the swagger definition:

* **if**  - If a condition evaluates to True, execute the corresponding flow.
* **switch** – Given a set of conditions, execute the flow corresponding to the
  first condition that evaluates to True.
* **operation-switch** – Given a set of conditions on the operation, select the
  first rule where the condition evaluates to True.
* **throw** – Throw an exception
* **invoke** – Retrieve a resource using HTTP or HTTPS. Currently supported HTTP
  request methods or verbs are
  `GET`, `HEAD`, `POST`, `PUT`, `PATCH`, `DELETE`, and `OPTIONS`.
* **javascript** – Execute a JavaScript program to manipulate or inspect the
  transaction
* **set-variable** – Set a variable in the context

# API Designer Toolkit

The API Designer toolkit provides a graphical interface for the Microgateway.
You can download it using NPM via `npm install -g apiconnect`. This toolkit can
be also be used for creating enterprise API definitions using IBM API Connect,
which provides features to create and manage APIs on an enterprise scale.

For more information, see https://developer.ibm.com/apiconnect/.

The API designer toolkit creates YAML files for the APIs definitions. These can
then be tested directly on the internal Microgateway (part of the API Designer
toolkit), or you can run them on an external Microgateway by moving the
underlying YAML files to the external Microgateway directory.

# The Microgateway Architecture

The Microgateway was developed with the goal of making something simple,
community-based, and that could easily be extended for anyone’s needs.

The Microgateway has the following characteristics:

* Built on Node Express and Loopback frameworks
* Processes Swagger API definitions via a middleware chain that identifies the
  API and executes the API definition.
* Contains a “datastore” that holds the data model of all API artifacts to be
  processed.
* Uses a flow-engine to process a variety of policies giving the API designer
  the ability to perform deep packet processing on the request or response.

A diagram of the Microgateway is shown below. Looking at the diagram, the flow
of a request is from left to right.  The Microgateway is a collection of
middleware components that process the request in order. Each middleware
component passes control to the next middleware until the processing of the
request is complete.  The **postflow** and **error-handler** middlewares work
together to return the results back to the client making the request.

![alt text][microgateway-components]

[microgateway-components]: https://github.com/strongloop/microgateway/blob/master/images/readme/MicroGatewayArchitecture2.png "Microgateway Component Structure"

The **urlrewrite** middleware simply modifies the prefix of the URL under
certain conditions.  For the most part, this is a passthrough.

The **context** middleware creates a scratchpad memory area known as the
context. The context is accessible by all middlewares and policies in the flow.
Any middleware or policy can add, modify or remove variables from the context.

Here are some of the context variables that are automatically populated:

![alt text][context-variables-part1]

[context-variables-part1]: https://github.com/strongloop/microgateway/blob/master/images/readme/ContextVariablesPart1.png "Microgateway Context Variables"

One object that is particularly important is the **message** object. The
**message** object contains the payload that was received from a request. For
example, if you add an **invoke** action, the results from that action will be
placed in the **message** object. At the end of the flow, the contents of the
**message** object will be returned back to the client.

Here are some other context variables:

![alt text][context-variables-part2]

[context-variables-part2]: https://github.com/strongloop/microgateway/blob/master/images/readme/ContextVariablesPart2.png "Microgateway Context Variables"

The **request** object is another important object. It holds all of the
information about the original request that was received by the Microgateway.
There are other objects that contain system information, plan information, and
general API information.

One important aspect of the context is that it is read-writable by policies and
middleware as they execute. Another important factor is that context variables
can be used as substitution parameters inside the conditional policies. This
allows you to write sophisticated logic flows, simply referencing them through
configuration.

The **analytics** middleware is used to connect to an external analytics engine.
It passes a series of variables from the context in a message to an external
collection device.

The preflow middleware accomplishes the following:

1.	Identifies the API (or Swagger definition) to execute.
2.	Performs security associated with that Swagger definition.
3.	Performs rate-limiting for the API.
4.	Creates the objects necessary for the assembly (IBM extensions to the
swagger definition) to be consumed by the flow-engine.

The flow-engine is a combinational processor that allows you to insert
sequential logic around a series of policies. The policies can perform any
operation to the request or response payload. They can be used to retrieve a
response from an API backend or examine the request for security or other needs.

The flow-engine is built as a vendor extension to the standard Swagger
specification. The policies that are referenced in the assembly must have a
pre-built policy. Each one of the policies is a Node.js module that provides the
processing of the message. Each policy also has a policy.yml file that defines
the set of properties and behavior of the policy. For examples, visit the
`policies` directory of the microgateway repository.


# Getting Started with the Microgateway
Following are the steps to install and run a stand alone microgateway

Step 1. Clone the microgateway repository
```
cd $HOME
git clone https://github.com/strongloop/microgateway.git
```

Step 2. Populate all of the necessary dependencies for the project
```
cd $HOME/microgateway
npm install
```

Step 3. Change current working directory to the root directory
```
cd $HOME/microgateway/
```

Step 4. Create a startup script that sets environment variables and starts up
the Microgateway. The script file is a simple node.js JavaScript file shown
below. Create this file in the `$HOME/microgateway/` directory.

Note:  The CONFIG_DIR is the folder containing the yaml files holding the API
definitions that you wish to enforce.
```
// sample.js
//
'use strict';

var mg = require('./lib/microgw');
var fs = require('fs');

// config dir
process.env.CONFIG_DIR = __dirname + '/definitions/myapp';
process.env.NODE_ENV = 'production';
mg.start(3000);
```

Step 5. Create a yaml file to define the API. Place the yaml file in the folder
identified by the `CONFIG_DIR` environment variable created in the startup script.
For this example, we are creating the file sample_1.0.0.yaml in the
`$HOME/microgateway/definitions/myapp` directory. Note that you can place
several yaml files in this directory and all will be pulled in and used by the
Microgateway.
```
# sample_1.0.0.yaml
#
info:
  version: 1.0.0
  title: sample
  description: sample laptop yaml
basePath: /sample
swagger: '2.0'
paths:
  /echo:
    get:
      responses:
        '200':
          description: 200 OK
x-ibm-configuration:
  assembly:
    execute:
      - javascript:
          title: write a small json object
          source: |
           message.body = { text : 'Hello World' };
schemes:
  - http
```

Step 6. From the root directory, execute the command to start the Microgateway.
```
cd $HOME/microgateway/
node sample.js
```

Step 7. Send a curl command to test the API. It should return the
`{“text”:”Hello World”}` JSON object.
```
curl http://localhost:3000/sample/echo
```

For more information on the internal specifics of the Microgateway, you may
want to look at the microgateway/test directory. All middleware components
have one or more test suites to exercise their interfaces and logic.

For more examples on how to add various policies to your API assembly, look in
the microgateway/test/definitions directory. There are several swagger files
that are used to test out the implementation.

# How to contribute to this Project

For information on contribuging to this project, please look at CONTRIBUTING.md
and CONDUCT.md as well as the LICENSE.txt file.




CHANGES.md
2019-11-05, Version 1.7.4
=========================

* Fix analytics data - app_id (Gary Tu)

2019-03-12, Version 1.7.3
=========================



2019-03-08, Version 1.7.2
=========================

 * Update datastore depenency (Gary Tu)


2019-03-06, Version 1.7.1
=========================

 * Update shrinkwrap (Gary Tu)


2019-02-22, Version 1.7.0
=========================

 * Add the shrinkwrap file (Gary Tu)

 * Add appName in analytics data (Gary Tu)

 * Exit mocha test promptly (Gary Tu)

 * Fix failed tests for node/dep version upgrade (Gary Tu)

 * Rebase with 1.6.x and update vulnerable deps (Gary Tu)

 * docs: add statement for supported HTTP verbs (biniam)

 * - package.json version up for datastore (Anton Korotkikh)

 * - downgrade to es5 (Anton Korotkikh)

 * readme (Anton Korotkikh)

 * cash fix (Anton Korotkikh)

 * env fixes (Anton Korotkikh)

 * env var fix (Anton Korotkikh)

 * - some changes for using cached datastore (Anton Korotkikh)

 * Double Step 4 (Ludovic)

 * Modified readme instructions to fix script typo (Ozair Sheikh)

 * Issues with Dockerizing nginx with microgateway (Ozair Sheikh)

 * test: avoid using port 7000 in tests (Ryan Graham)

 * Add config for building on Travis (Ryan Graham)

 * fix platform-specific compression test failure (Ryan Graham)

 * Fix DBCS issue (#63) (Jeremy Geddes)

 * use conventional license format and annotation (Ryan Graham)

 * Fix the shell scipt (Raymond Feng)

 * Link to our developer page (Tony Ffrench)

 * add back redis and ip modules (Anthony Ffrench)

 * basic nginx TLS termination (Anthony Ffrench)

 * cleanup the dockerfile, rm unused deps (Anthony Ffrench)

 * add ab command example (Clement)

 * add jmeter example (Clement)

 * update performance readme (Clement)

 * fix linter failures (Anthony Ffrench)

 * update products_template (cwsu)

 * multiple subscription and credentials (cwsu)

 * change to new rate-limit format (cwsu)

 * fix path problem (cwsu)

 * ratelmit config (cwsu)

 * create tools folder and draft a md file (cwsu)

 * perf test framework security part (Clement)

 * add security settings (Clement)

 * first draft of perf test (Clement)

 * Fix the image references (Tony Ffrench)

 * README.md formating cleanup (Anthony Ffrench)

 * modernize the eslint config, engine>4 (Anthony Ffrench)

 * pick a better invalid hostname (Anthony Ffrench)

 * revert .gitignore (Clement)

 * remove genkeycert script (Clement)

 * change path from develop to master (Clement)

 * fix the image link in README.md (Clement)

 * restore key and cert (Clement)

 * Change sensitive domain to example.com (Clement)

 * correct info (Clement)

 * add Ozair to maintainers list (Clement)

 * add/modify docs for open source (Clement)

 * Spelling corrections (Timothy Smith)

 * Updated README.md for Open Source (Timothy Smith)

 * modify copyright for open source (Clement)

 * merge in fixes and changes for gateway director (#28) (Jeremy Geddes)

 * change dependancy to user reference for now (Jeremy R Geddes)

 * move from index.js to datastore.js (Jeremy R Geddes)

 * format issue (Jeremy Geddes)

 * add note about key/gen changes (Jeremy Geddes)

 * Remove key/cert and autogen in postinstall phase (#18) (Gary Tu)

 * add message change note.. (Jeremy Geddes)

 * Issue/error msgs (#20) (Daniel Badt)

 * Datastore repo split (#26) (Jeremy Geddes)

 * Rebase off master, missed merges (Krishna Raman)

 * Ensure config/default dir is created (Daniel Badt)

 * Add timeout to failing test. (Rick Curtis)

 * Small change to trigger ci. (Rick Curtis)

 * Update to use config/project modules. (Rick Curtis)

 * Fix git url (Krishna Raman)

 * Add gateways section (Chris Bygrave)

 * Tidy up whitespace (Chris Bygrave)

 * Add a handlebars policy & tests (Chris Bygrave)


2016-12-06, Version 1.6.1
=========================

 * Ensure config/default dir is created (Daniel Badt)

 * Fix git url (Krishna Raman)

 * Add timeout to failing test. (Rick Curtis)

 * Update to use config/project modules. (Rick Curtis)


2016-10-13, Version 1.5.6
=========================

 * Fix writeFileSync calls (Daniel Badt)

 * File chg should not trigger timer (Daniel Badt)

 * Misc. fixes for webhooks (Daniel Badt)

 * Update the datastore/README.md (juangmj)


2016-10-11, Version 1.5.5
=========================

 * this is the current template for readme files (Om Goeckermann)

 * added new information - entries for default application and plan (Om Goeckermann)

 * fix timing issue of test case (Clement)

 * fix indetation problem on test data (Clement)

 * monitor local files for laptop experience (Clement)

 * Fix issue with webhooks unsubscribe (Daniel Badt)


2016-10-06, Version 1.5.4
=========================



2016-11-23, Version 1.6.0
=========================

 * Fix writeFileSync calls (Daniel Badt)

 * File chg should not trigger timer (Daniel Badt)

 * Misc. fixes for webhooks (Daniel Badt)

 * Update the datastore/README.md (juangmj)

 * this is the current template for readme files (Om Goeckermann)

 * added new information - entries for default application and plan (Om Goeckermann)

 * fix timing issue of test case (Clement)

 * fix indetation problem on test data (Clement)

 * monitor local files for laptop experience (Clement)

 * Fix issue with webhooks unsubscribe (Daniel Badt)

 * Update README.md (Om Goeckermann)

 * Fix expected unsubscribe response code (Daniel Badt)

 * Add webhooks unsubscribe (Daniel Badt)

 * Address lint failures (Jon Palgon)

 * Some fixes for webhooks (Daniel Badt)

 * Initial changes for MGW webhooks (Daniel Badt)

 * Change space array names to plural (Daniel Badt)

 * uGW support for spaces (Daniel Badt)

 * Initialize the test-app-cid-sec to true (juangmj)

 * Allow test-app to call OAuth2 API (juangmj)

 * Check arbitrary apikeys before fixed (Daniel Badt)

 * Allow developers to configure plan and rate-limit (juangmj)

 * Should not reject GET requests with payload (juangmj)


utils/utils.js
// Copyright IBM Corp. 2016. All Rights Reserved.
// Node module: microgateway
// US Government Users Restricted Rights - Use, duplication or disclosure
// restricted by GSA ADP Schedule Contract with IBM Corp.

var env = require('./environment');
var path = require('path');
var fs = require('fs');
var Promise = require('bluebird');
var log = require('apiconnect-cli-logger/logger.js')
                  .child({ loc: 'microgateway:utils' });
var crypto = require('crypto');
var url = require('url');
var request = require('request');
var constants = require('constants');
var wlpnPassword = require('wlpn-password');

var version = require('../package.json').version;

exports.getTLSConfigSync = function() {
  var rev;
  var cfg = process.env[env.TLS_SERVER_CONFIG] ?
      //the env value should be a relative path to parent directory
      path.resolve(__dirname, '..', process.env[env.TLS_SERVER_CONFIG]) :
      path.resolve(__dirname, '..', 'config', 'defaultTLS.json');

  try {
    rev = JSON.parse(fs.readFileSync(cfg));
    var baseDir = path.dirname(cfg); //base dir, used for relative path
    var props = [ 'pfx', 'key', 'cert', 'ca', 'dhparam', 'ticketKeys', 'passphrase' ];
    for (var index = 0, length = props.length; index < length; index++) {
      var propName = props[index];
      if (rev[propName] instanceof Array) { // ca is capable of being array
        var values = rev[propName];
        var newValues = [];
        for (var valueIndex = 0, valueLength = values.length;
            valueIndex < valueLength; valueIndex++) {
          try {
            newValues.push(fs.readFileSync(path.resolve(baseDir, values[valueIndex])));
          } catch (e) {
            newValues.push(values[valueIndex]);
          }
        }
        rev[propName] = newValues;
      } else if (rev[propName]) {
        var filename = rev[propName];
        var property;
        if (filename.indexOf(':') !== -1) {
          var array = filename.split(':');
          filename = array[0];
          property = array[1];
        }
        try {
          var contents = fs.readFileSync(path.resolve(baseDir, filename));
          if (property) {
            var parsedFile = JSON.parse(contents);
            rev[propName] = parsedFile[property];
          } else {
            rev[propName] = contents;
          }
        } catch (e) {
          // do nothing
        }
      }
    }
    if (rev.passphrase) {
      rev.passphrase = wlpnPassword.decode(rev.passphrase);
    }
  } catch (e) {
    log.error(e);
  }

  return rev || {};
};


/**
 * sign msg with the given key and alg
 */
function signMsg(msg, key, alg) {
  var signer = crypto.createSign(alg);
  signer.update(msg);
  return signer.sign(key);
}

/**
 * hash the msg with the msg and alg
 */
function hashMsg(msg, alg) {
  var hash = crypto.createHash(alg);
  hash.update(msg);
  return hash.digest();
}

/**
 * This function decrypts and parses an encrypted response body sent by APIM
 * The body must be in the following format:
 *
 * {
 *   "key": "base64(encrypted_with_public_key(aes_256_symmetric_key))"
 *   "cipher": "base64(encrypted_with_aes_256_key(json_payload_as_string))"
 * }
 * @param body
 * @param private_key
 *
 */
function decryptAPIMResponse(body, private_key) {

  if (!body.key || !body.cipher) {
    throw new Error('bad handshake response from APIm');
  }

  var key = crypto.privateDecrypt(
    { key: private_key,
      padding: constants.RSA_PKCS1_PADDING },
    new Buffer(body.key, 'base64')
  );

  var iv = new Buffer(16);
  iv.fill(0);
  var decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
  var plainText = decipher.update(body.cipher, 'base64', 'utf8');
  plainText += decipher.final('utf8');

  log.debug('handshake response payload:', plainText);
  return JSON.parse(plainText);
}

/**
 * Compute the signature headers "date", "digest", and "authorization" headers
 * according to IETF I-D draft-cavage-http-signatures-05 using rsa-sha256 algorithm
 *
 * If the `date` header already exists in the input, it's used as-is
 * If the `digest` header already exists in the input, it's used as-is (which means that body is ignored)
 *
 *
 * @param body (String): Message body (ignored if there is already a digest header)
 * @param headers (Object): Contains the existing list of headers
 * @param keyId (String): Identifier for the private key, ends up in the "keyId" param of the authorization header
 * @param key (String): RSA Private key to be used for the signature
 * @returns {*}
 */
function addSignatureHeaders(body, headers, keyId, key) {
  if (!headers) {
    headers = {};
  }

  if (!headers.date) {
    headers.date = (new Date()).toUTCString();
  }

  if (!headers.digest) {
    headers.digest = 'SHA256=' + hashMsg(JSON.stringify(body), 'sha256')
        .toString('base64');
  }


  var combine = function(names, headers) {
    var parts = [];
    names.forEach(function(e) {
      parts.push(e + ': ' + headers[e]);
    });
    return parts.join('\n');
  };

  headers.authorization = 'Signature ' +
    'keyId="' + keyId + '", ' +
    'headers="date digest", ' +
    'algorithm="rsa-sha256", ' +
    'signature="' +
    signMsg(combine([ 'date', 'digest' ], headers), key, 'RSA-SHA256')
      .toString('base64') + '"';

  return headers;
}


/**
 * Attempt to handshake from APIm server
 * @param {Object} apim - configuration pointing to APIm server
 * @param {string} privKey - private key to be used for handshake
 * @param {callback} doneCB - done callback
 */
exports.handshakeWithAPIm = function(apim, privKey, doneCB) {
  log.debug('handshakeWithAPIm entry');

  if (privKey instanceof Function) {
    doneCB = privKey;
    try {
      privKey = fs.readFileSync(path.resolve(__dirname, '..', env.KEYNAME));
    } catch (e) {
      doneCB(new Error('can not load default private key'));
      return;
    }
  }

  new Promise(function(resolve, reject) {
    var body = JSON.stringify({ gatewayVersion: version });

    var headers = { 'content-type': 'application/json' };

    addSignatureHeaders(body, headers,
        'micro-gw-catalog/' + apim.catalog, privKey);

    if (log.debug()) {
      log.debug(JSON.stringify(headers, null, 2));
    }

    var apimHandshakeUrlObj = {
      protocol: 'https',
      hostname: apim.host,
      port: apim.port,
      pathname: '/v1/catalogs/' + apim.catalog + '/handshake/' };

    var targetURL = url.format(apimHandshakeUrlObj);

    var options = {
      url: targetURL,
      method: 'POST',
      json: body,
      headers: headers,
      //FIXME: need to eventually remove this
      agentOptions: { rejectUnauthorized: false } };

    request(options,
      function(err, res, body) {
        if (err) {
          reject(new Error(
              'Failed to communicate with %s: %s ' + targetURL + err));
          return;
        }

        log.debug('statusCode: ' + res.statusCode);
        if (res.statusCode !== 200) {
          reject(new Error(targetURL + ' failed with: ' + res.statusCode));
          return;
        }

        try {
          var json = decryptAPIMResponse(body, privKey);

          if (!json || !json.microGateway) {
            throw new Error(
                targetURL + ' response did not contain "microGateway" section');
          }
          //{cert: cert, key:key, clientID:clientID}
          resolve(json.microGateway);
        } catch (e) {
          reject(e);
        }

      }
    ); //request end

  })
  .then(function(result) {
    log.info('Successful handshake with API Connect server');
    doneCB(undefined, { clientID: result.clientID });
    log.debug('handshakeWithAPIm exit');
  })
  .catch(function(error) {
    log.error('Unsuccessful handshake with API Connect server');
    doneCB(error);
  });

};
